### ISSUE TITLE
📝 Docs: Explain *why* each package should be replaced

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [es-tooling/module-replacements#187]
I personally am bought into the idea of dropping bloat from dependencies 😄. But, for folks who're new to this, their first introduction might be a `modules/*.md` page in this repo. Many of those pages don't really do much to explain *why* the package in question should be replaced.

For example, `glob`:

```plaintext
/Users/josh/repos/mocha/lib/cli/lookup-files.js
  10:12  error  "glob" should be replaced with an alternative package. Read more here: https://github.com/es-tooling/module-replacements/blob/main/docs/modules/glob.md  depend/ban-dependencies
```

https://github.com/es-tooling/module-replacements/blob/main/docs/modules/glob.md just says:

https://github.com/es-tooling/module-replacements/blob/5560f1787853cbe1a0b359be719efe0430ebe867/docs/modules/glob.md?plain=1#L3

I as a theoretical new user am actually told: what's so bad about `glob`?

Suggestion: for each page, including at least `glob`, mandate at least a sentence or two explaining the problems (perf, size, etc.) of the package in question?

**Comment by lukekarrys** [comment]
I think this a really good idea.

Related to this, I also think there’s room for a section like “Why you might *not* want to migrate away from X” for some modules, especially ones that get replaced by core. 

I have a few specific examples from my experience that I think could help users make more informed decisions. The goal should be to present information that will increase confidence even if the decision is still to migrate away.

Thinking from a similar perspective of a theoretical new user, they should read these and think “my code is not doing any of that <insert specific example>, this is safe to replace”

Here are some rough examples from my experience:

- `fs-extra`: This library also applies `graceful-fs` if your code does not need to be resilient to `EMFILE` errors then it is safe to migrate away. If it does, consider installing and applying `graceful-fs` yourself.
- `rimraf`: One of the primary use cases to rimraf is to be resilient to transient `EPERM` (and similar) errors on Windows. These tend to only happen when removing many files in rapid succession on resource constrained machines. If this is a concern of your code, consider handling these errors manually or sticking with rimraf [^1].

[^1]: I removed most of the uses of rimraf from npm so it could be dropped as a dependency in favor of core fs methods. I also have a WIP PR for rimraf that increases resiliency on Windows where I test by deleting files in CI in a loop for 30+ minutes.

---

Please minimize if this is off-topic. I can also open a new issue if a further discussion around this is warranted.

**Comment by 43081j** [comment]
Fair point! 

We haven't yet put much effort into the explanations in each doc. It would be good to expand a little, although some literally are just because they're smaller and/or faster

Mentioning situations where you may not want to migrate is a good idea too

**Comment by lukekarrys** [comment]
> It would be good to expand a little, although some literally are just because they're smaller and/or faster

I singled out the ones that are
mostly replaced by core because those are more likely to be nuanced due to the tradeoffs of what gets implemented in core vs userland.

For ones where there is a new userland dropin(-ish) replacement, I don't think this section would be necessary.

Happy to create a few separate PRs for modules I have experience with if that would be helpful.

**Comment by 43081j** [comment]
sure if you want to have a stab at it, maybe that'll at least give us a PR to discuss. then whatever structure we settle on in that can become the template

**Comment by gameroman** [comment]
> we don't do this yet for each replacement, but have started doing it in new documented replacements.
> 
> so i think this is still open until we add brief reasoning to all existing documented replacement docs.
> 
> for anyone thinking of picking that up - please don't do it in individual PRs as it will be a lot and will be too many to review individually.

I've checked and most documented replacements do. Do you have an example of one that does not? So that I can understand what is meant by "brief reasoning"

**Comment by 43081j** [comment]
quite a lot of them mention that they can be replaced by "more performant packages", but no explicit reasoning to back that up, etc.

i think this one needs some deeper thought before we can action it. i'll bring it up in discord
